### PR TITLE
GH-41675: [Packaging][MATLAB] Add crossbow job to package MATLAB interface on macos-14

### PR DESCRIPTION
**Pull request by sgilmore10**
### Rationale for this change

In #41592, we added a CI job to build and test the MATLAB interface on `macos-14`, which is ARM-based. We currently only package the interface on `macos-12`, which is Intel-based. 

We should add a crossbow job to package the MATLAB interface on `macos-14` so that we package the interface for both Intel-based and ARM-based macoS. 

### What changes are included in this PR?

1. Parameterized the `macos` job in the MATLAB crossbow workflow file to run on both `macos-12` and `macos-14`.
2. Added the bash script `dev/tasks/matlab/rename_macos_dynamic_libraries.sh`. This script is used to uniquify the shared library names generated for the Intel/AMD-based macOS and ARM-based macOS interface installations. This is required because the crossbow job generates one "monolithic" MLTBX file that contains all shared libraries for all platforms. See the [comment](https://github.com/apache/arrow/pull/41677/files#diff-99731016fc39dd96d2d239cf9cc132a8dacb9569c49bea4a64d69911b4dcc8c4R82) at the beginning of `dev/tasks/matlab/rename_macos_dynamic_libraries.sh` for a more in-depth explanation. 

### Are these changes tested?

1. Installed the MATLAB Arrow Interface on both Intel-based and ARM-based macOS machines using the MLTBX file generated by [this crossbow job](https://github.com/ursacomputing/crossbow/actions/runs/9274264286/job/25516244719). All unit tests passed.

### Are there any user-facing changes?

Users on ARM-based macOS machines will be able to install the MATLAB Arrow Interface via a "one-click" install 
workflow using the MLTBX file.

* GitHub Issue: #41675

### CODE REVIEW

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

File: dev/tasks/matlab/github.yml [hunk 1]
```diff
@@ -73,27 +79,18 @@ jobs:
...
-           pushd arrow/matlab/install/arrow_matlab/+libmexclass/+proxy
-           SYMLINK_ARROW_LIB="$(find . -name 'libarrow.*.dylib' -type l | xargs basename)"
-           REGULAR_ARROW_LIB="$(echo libarrow.*.*.dylib)"
-           echo "SYMLINK_ARROW_LIB = ${SYMLINK_ARROW_LIB}"
-           echo "REGULAR_ARROW_LIB = ${REGULAR_ARROW_LIB}"
-           install_name_tool -change @rpath/$SYMLINK_ARROW_LIB @rpath/$REGULAR_ARROW_LIB libarrowproxy.dylib
-           popd
+          arrow/dev/tasks/matlab/rename_macos_dynamic_libraries.sh \
+          arrow/matlab/install/arrow_matlab/+libmexclass/+proxy \
+          {{ '${{ matrix.platform.architecture-suffix }}' }}
```
Comment (kou) [inline dev/tasks/matlab/github.yml:86]:
```suggestion
            arrow/matlab/install/arrow_matlab/+libmexclass/+proxy \
            {{ '${{ matrix.platform.architecture-suffix }}' }}
```

File: dev/tasks/matlab/github.yml [hunk 2]
```diff
@@ -59,8 +59,14 @@ jobs:
           path: matlab-arrow-ubuntu.tar.gz
 
   macos:
-    name: AMD64 macOS 12 MATLAB
-    runs-on: macos-12
+    {% set arch = '${{ matrix.platform.architecture}}' %}
+    name: {{ arch }} macOS MATLAB
```
Comment (kou) [inline dev/tasks/matlab/github.yml:63]: It seems that `arch` is only used here. Can we simplify this?

```suggestion
    name: {{ '${{ matrix.platform.architecture}}' }} macOS MATLAB
```

File: dev/tasks/matlab/rename_macos_dynamic_libraries.sh [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  IS_ARM64=1
+elif [ "$ARCH" == "x64" ]; then
+  IS_ARM64=0
+else
+  echo "<arch> must be arm64 or x64"
+  exit 1
+fi
+
+ORIG_DIR=$(pwd)
+
```
Comment (kou) [inline dev/tasks/matlab/rename_macos_dynamic_libraries.sh:113]: We don't need this. If we want to return to the original directory, we can use `pushd`/`popd`:

```shell
pushd ${DYLIB_DIR}
# ...
popd
```

File: dev/tasks/matlab/rename_macos_dynamic_libraries.sh [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+# Change install name of dependent shared library libmexclass.dylib to libmexclass_(arm64|x64).*.*.*.dylib libarrowproxy_(arm64|x64).dylib
+install_name_tool -change @rpath/${LIBMEXCLASS_DYLIB} @rpath/${LIBMEXCLASS_ARCH_DYLIB} ${LIBARROWPROXY_ARCH_DYLIB}
+
+# Change install name of dependent shared library libmexclass.dylib to libmexclass_(arm64|x64).dylib in gateway.(mexmaca64|mexmaci64)
+install_name_tool -change @rpath/${LIBMEXCLASS_DYLIB} @rpath/${LIBMEXCLASS_ARCH_DYLIB} ${MEX_GATEWAY}
+
+# Change install name of dependent shared library libarrowproxy.dylib to libarrowproxy_(arm64|x64).dylib in gateway.(mexmaca64|mexmaci64)
+install_name_tool -change @rpath/${LIBARROWPROXY_DYLIB} @rpath/${LIBARROWPROXY_ARCH_DYLIB} ${MEX_GATEWAY}
+
+cd ${ORIG_DIR}
```
Comment (kou) [inline dev/tasks/matlab/rename_macos_dynamic_libraries.sh:164]:
```suggestion
```

**Review by kou — APPROVED** [review]
+1

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

File: dev/tasks/matlab/github.yml (hunk 2 above)
Comment (sgilmore10) [reply dev/tasks/matlab/github.yml]: Will do!

File: dev/tasks/matlab/rename_macos_dynamic_libraries.sh (hunk 4 above)
Comment (sgilmore10) [reply dev/tasks/matlab/rename_macos_dynamic_libraries.sh]: Done!

**Comment by sgilmore10** [comment]
@github-actions crossbow submit matlab

**Comment by sgilmore10** [comment]
Re-ran the crossbow job with the new changes. Everything passed.

**Pull request merged**